### ISSUE TITLE
`hostPort` pod restarts result in proxy connection failures

### ISSUE THREAD

**Issue by jault-figure** [linkerd/linkerd2#7589]
### What is the issue?

I have Datadog running as a daemonset in my cluster. Each Datadog agent has a container that [uses a `hostPort`](https://github.com/DataDog/datadog-agent/blob/main/Dockerfiles/manifests/all-containers/daemonset.yaml#L157) to accept tracing data from applications. Applications send data to the _host_ IP address (not a service name) and port 8126. I believe Datadog recommends that approach, or at least part of the reason, is to properly tag these metrics with the host the application is running on so you're able to determine if a single host is having problems.

When a datadog agent redeploys (the pod is deleted & recreated), other applications on that same host are unable to reconnect even after it comes back up healthy. The Linkerd proxy logs show connection failures, no route to host errors, etc (shown below). I would expect that once the Datadog agent comes back up, the Linkerd proxy is able to successfully make connections again.

The only way I've found so far to fix this is by redeploying the application, or skipping port 8126 via the `config.linkerd.io/skip-outbound-ports: "8126"` annotation. However, the annotation doesn't seem like a step that should be needed long term.

### How can it be reproduced?

1. Deploy datadog according to [their documentation](https://github.com/DataDog/datadog-agent/blob/main/Dockerfiles/manifests/all-containers/daemonset.yaml). Configure it to enable APM.
2. Inject the host IP as the datadog agent address [described here](https://docs.datadoghq.com/agent/kubernetes/apm/?tab=daemonset#configure-your-application-pods-in-order-to-communicate-with-the-datadog-agent)
```
        env:
          - name: DD_AGENT_HOST
            valueFrom:
              fieldRef:
                fieldPath: status.hostIP
```
3. Configure the application to send tracing data to datadog. For Java apps, [these steps](https://docs.datadoghq.com/tracing/setup_overview/setup/java/?tab=containers) can be following.
4. Delete the datadog agent running on the same host as the application.
5. Follow the linkerd-proxy logs on the application to see the errors.

### Logs, error output, etc

Linkerd proxy logs the following errors continuously.
```
[361302.280329s]  INFO ThreadId(01) outbound:server{orig_dst=192.168.0.86:8126}:rescue{client.addr=10.0.45.165:56702}: linkerd_app_core::errors::respond: Request failed error=HTTP Server service in fail-fast
[361303.383634s]  WARN ThreadId(01) outbound:server{orig_dst=192.168.0.86:8126}: linkerd_reconnect: Failed to connect error=connect timed out after 1s
[361303.878323s]  WARN ThreadId(01) outbound:server{orig_dst=192.168.0.86:8126}: linkerd_reconnect: Failed to connect error=No route to host (os error 113)
[361305.082594s]  WARN ThreadId(01) outbound:server{orig_dst=192.168.0.86:8126}: linkerd_reconnect: Failed to connect error=connect timed out after 1s
[361305.383127s]  WARN ThreadId(01) outbound:server{orig_dst=192.168.0.86:8126}: linkerd_stack::failfast: HTTP Server entering failfast after 3s
```

The application is showing that it receives a 504 response:
```
[dd.trace 2021-12-27 20:52:48:619 +0000] [dd-trace-processor] WARN datadog.trace.agent.common.writer.ddagent.DDAgentApi - Error while sending 1 (size=2KB) traces to the DD agent. Total: 1118154, Received: 1118154, Sent: 19997, Failed: 1098157. Status: 504 Gateway Timeout (Will not log errors for 5 minutes)
[dd.trace 2021-12-27 20:57:48:684 +0000] [dd-trace-processor] WARN datadog.trace.agent.common.writer.ddagent.DDAgentApi - Error while sending 2 (size=2KB) traces to the DD agent. Total: 1119088, Received: 1119088, Sent: 19997, Failed: 1099091. Status: 504 Gateway Timeout (Will not log errors for 5 minutes)
[dd.trace 2021-12-27 21:02:51:345 +0000] [dd-trace-processor] WARN datadog.trace.agent.common.writer.ddagent.DDAgentApi - Error while sending 14 (size=14KB) traces to the DD agent. Total: 1120028, Received: 1120028, Sent: 19997, Failed: 1100031. Status: 504 Gateway Timeout (Will not log errors for 5 minutes)
```

### output of `linkerd check -o short`

```
Linkerd core checks
===================

linkerd-identity
----------------
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2022-02-02T12:37:12Z
    see https://linkerd.io/2.11/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints

linkerd-webhooks-and-apisvc-tls
-------------------------------
‼ proxy-injector cert is valid for at least 60 days
    certificate will expire on 2022-01-12T03:37:11Z
    see https://linkerd.io/2.11/checks/#l5d-proxy-injector-webhook-cert-not-expiring-soon for hints
‼ sp-validator cert is valid for at least 60 days
    certificate will expire on 2022-01-12T03:37:11Z
    see https://linkerd.io/2.11/checks/#l5d-sp-validator-webhook-cert-not-expiring-soon for hints
‼ policy-validator cert is valid for at least 60 days
    certificate will expire on 2022-01-12T03:37:13Z
    see https://linkerd.io/2.11/checks/#l5d-policy-validator-webhook-cert-not-expiring-soon for hints

Status check results are √

Linkerd extensions checks
=========================


Status check results are √

linkerd-viz
-----------
‼ tap API server cert is valid for at least 60 days
    certificate will expire on 2022-01-12T04:04:33Z
    see https://linkerd.io/2.11/checks/#l5d-tap-cert-not-expiring-soon for hints

Status check results are √
```

Just an FYI for the output above, certificates are automatically managed by cert manager with a pretty short expiration in this environment.

### Environment

Cluster:
GKE cluster running version 1.19.14

Hosts:
Hosts are running Container-Optimized OS with Docker

Linkerd version:
```
$ linkerd version
Client version: stable-2.11.1
Server version: stable-2.11.1
```

### Additional context

* The application was left in this state for multiple hours without seeing any recovery, so it doesn't seem to be a timing issue
* Both the application and the Datadog agent have been injected. But since the application is sending data to the host's IP address, I don't believe the proxy would use mTLS.
* I enabled debug logging on the control plane components and have verified that the endpoints are updated correctly in the destination service when datadog redeploys. But those endpoints only contain the pod IP, not the host IP, which is expected. Since this traffic goes to the host IP I wouldn't expect this to matter much, but I checked it out anyways 😄 
* I initially thought this was related to the `hostPort` that Datadog uses, but a similar looking issue was reported in the [linkerd slack](https://linkerd.slack.com/archives/C89RTCWJF/p1640347423088800). That issue reported in Slack does not involve a `hostPort` so I figured opening an issue to get some more eyes would be helpful!

### Would you like to work on fixing this bug?

no

**Comment by jault3** [comment]
It looks like this issue was assigned. @kleimkuhler do you know if this is still planning to be included in the 2.12.0 milestone? Was hoping to remove the stale label if it is

**Comment by kleimkuhler** [comment]
It'd be unfortunate to see this stale out, but we've been unable to reproduce this. If anyone has run into this and can provide something reproducible then we should be able to start looking into this again.

**Comment by alpeb** [comment]
#9819 fixed an issue with hostPorts that got shipped in stable-2.12.3. That fixed an issue resulting in 502 errors on the client-side unlike the 504 reported here, but it's very much likely the same underlying issue. So closing this one out. Please report back if you still experience this issue (or if upgrading fixed it for you!).
